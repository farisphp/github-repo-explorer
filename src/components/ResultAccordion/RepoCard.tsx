import { FaStar } from 'react-icons/fa'
import { RestEndpointMethodTypes } from '@octokit/rest'
import { formatDistanceToNow } from 'date-fns'

function RepoCard({
  data
}: {
  data: RestEndpointMethodTypes['repos']['listForUser']['response']['data'][0]
}) {
  return (
    <div key={data.html_url} className="rounded-lg border p-4">
      <div className="flex items-center space-x-2">
        <a href={data.html_url}>
          <h3 className="w-full border-b text-base font-semibold hover:border-black">
            {data.name}
          </h3>
        </a>

        <div className="flex items-center">
          <FaStar />
          <span className="ml-1">{data.stargazers_count}</span>
        </div>
        {data.language ? (
          <span className="rounded-md border p-1 text-sm font-light">
            {data.language}
          </span>
        ) : null}
      </div>
      {data.updated_at ? (
        <span className="mt-1 block">
          Last update {formatDistanceToNow(new Date(data.updated_at))}
        </span>
      ) : null}

      <p className="mt-2">{data.description}</p>
    </div>
  )
}

export default RepoCard
