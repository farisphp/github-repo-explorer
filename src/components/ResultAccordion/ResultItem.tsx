import { useEffect } from 'react'
import { useFindRepoByUser } from 'hooks/repo'
import Spinner from 'components/Spinner'
import RepoCard from './RepoCard'

function ResultItem({ username }: { username: string }) {
  const { result, isLoading, findRepoByUser, error } = useFindRepoByUser()

  useEffect(() => {
    if (username) {
      findRepoByUser(username)
    }
  }, [username, findRepoByUser])

  return (
    <div className="space-y-2">
      {error?.message ? (
        <span className="block text-red-600">{error?.message}</span>
      ) : null}

      {isLoading ? <Spinner /> : null}
      {!isLoading && result?.data && result?.data?.length < 1
        ? 'No repository found'
        : null}
      {result?.data?.map((repo) => (
        <RepoCard key={repo.html_url} data={repo} />
      ))}
    </div>
  )
}

export default ResultItem
