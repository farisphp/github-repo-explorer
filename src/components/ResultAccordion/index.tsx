import {
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionTrigger
} from 'components/Accordion'
import ResultItem from './ResultItem'
import { RestEndpointMethodTypes } from '@octokit/rest'

function ResultAccordion({
  result
}: {
  result: RestEndpointMethodTypes['search']['users']['response'] | undefined
}) {
  return (
    <Accordion type="single" collapsible className="w-full">
      {result?.data?.items
        ? result?.data?.items?.map((item, idx) => {
            return (
              <AccordionItem key={`github-users-${idx}`} value={item.login}>
                <AccordionTrigger>
                  <div className="flex items-center">
                    {item.login}{' '}
                    <img
                      src={item.avatar_url}
                      className="ml-1 h-6 w-6"
                      alt=""
                    />
                  </div>
                </AccordionTrigger>
                <AccordionContent>
                  <ResultItem username={item.login} />
                </AccordionContent>
              </AccordionItem>
            )
          })
        : null}
    </Accordion>
  )
}

export default ResultAccordion
