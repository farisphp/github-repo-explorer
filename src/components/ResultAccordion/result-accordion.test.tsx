import { render, screen } from '@testing-library/react'
import ResultAccordion from '.'
import RepoCard from './RepoCard'
import ResultItem from './ResultItem'
import { RestEndpointMethodTypes } from '@octokit/rest'
import { Mocked, vi } from 'vitest'

const repoData = {
  id: 654181138,
  node_id: 'R_kgDOJv4DEg',
  name: 'github-repo-explorer',
  full_name: 'farisphp/github-repo-explorer',
  html_url: 'https://github.com/farisphp/github-repo-explorer',
  description: 'Find a user and their public repositories by their username.',
  created_at: '2023-06-15T14:53:21Z',
  updated_at: '2023-06-15T15:38:46Z',
  pushed_at: '2023-06-15T15:38:15Z',
  stargazers_count: 0,
  language: 'TypeScript'
}

const resultData = {
  total_count: 1,
  incomplete_results: false,
  items: [
    {
      login: 'farisphp',
      avatar_url: 'https://avatars.githubusercontent.com/u/22131960?v=4'
    }
  ]
}

vi.mock('hooks/repo', () => {
  const mockUseFindRepoByUser = vi.fn(() => ({
    result: {
      data: [
        repoData
      ] as unknown as RestEndpointMethodTypes['repos']['listForUser']['response']
    },
    isLoading: false,
    findRepoByUser: async (username: string) => {
      console.log()
    },
    error: undefined
  }))
  // Client.prototype.connect = vi.fn()
  // Client.prototype.query = vi.fn()
  // Client.prototype.end = vi.fn()

  return { useFindRepoByUser: mockUseFindRepoByUser }
})

describe('<ResultAccordion />', () => {
  it('should render result accordion correctly', () => {
    const { container } = render(
      <ResultAccordion
        result={
          {
            data: resultData
          } as RestEndpointMethodTypes['search']['users']['response']
        }
      />
    )

    const title = screen.getByRole('heading', { name: 'farisphp', level: 3 })
    expect(title).toBeInTheDocument()

    const buttonTrigger = screen.getByRole('button', {
      name: 'farisphp'
    })
    expect(buttonTrigger).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})

describe('<ResultItem />', () => {
  it('should render result item correctly', () => {
    const { container } = render(<ResultItem username="farisphp" />)

    const repoLink = screen.getByRole('link')
    expect(repoLink).toBeInTheDocument()
    expect(repoLink).toHaveAttribute('href', repoData.html_url)

    const repoTitle = screen.getByRole('heading', {
      name: 'github-repo-explorer',
      level: 3
    })
    expect(repoTitle).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})

describe('<RepoCard />', () => {
  it('should render repo card correctly', () => {
    const { container } = render(
      <RepoCard
        data={
          repoData as RestEndpointMethodTypes['repos']['listForUser']['response']['data'][0]
        }
      />
    )

    const repoLink = screen.getByRole('link')
    expect(repoLink).toBeInTheDocument()
    expect(repoLink).toHaveAttribute('href', repoData.html_url)

    const repoTitle = screen.getByRole('heading', {
      name: 'github-repo-explorer',
      level: 3
    })
    expect(repoTitle).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
