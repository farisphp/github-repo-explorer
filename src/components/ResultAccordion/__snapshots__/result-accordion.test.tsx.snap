// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RepoCard /> > should render repo card correctly 1`] = `
<div
  class="rounded-lg border p-4"
>
  <div
    class="flex items-center space-x-2"
  >
    <a
      href="https://github.com/farisphp/github-repo-explorer"
    >
      <h3
        class="w-full border-b text-base font-semibold hover:border-black"
      >
        github-repo-explorer
      </h3>
    </a>
    <div
      class="flex items-center"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 576 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
        />
      </svg>
      <span
        class="ml-1"
      >
        0
      </span>
    </div>
    <span
      class="rounded-md border p-1 text-sm font-light"
    >
      TypeScript
    </span>
  </div>
  <span
    class="mt-1 block"
  >
    Last update 
    about 2 hours
  </span>
  <p
    class="mt-2"
  >
    Find a user and their public repositories by their username.
  </p>
</div>
`;

exports[`<ResultAccordion /> > should render result accordion correctly 1`] = `
<div
  class="w-full"
  data-orientation="vertical"
>
  <div
    class="border-b"
    data-orientation="vertical"
    data-state="closed"
  >
    <h3
      class="flex"
      data-orientation="vertical"
      data-state="closed"
    >
      <button
        aria-controls="radix-:r1:"
        aria-expanded="false"
        class="flex flex-1 items-center justify-between py-4 font-medium transition-all hover:underline [&[data-state=open]>svg]:rotate-180"
        data-orientation="vertical"
        data-radix-collection-item=""
        data-state="closed"
        id="radix-:r0:"
        type="button"
      >
        <div
          class="flex items-center"
        >
          farisphp
           
          <img
            alt=""
            class="ml-1 h-6 w-6"
            src="https://avatars.githubusercontent.com/u/22131960?v=4"
          />
        </div>
        <svg
          class="h-4 w-4 transition-transform duration-200"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
    </h3>
    <div
      aria-labelledby="radix-:r0:"
      class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
      data-orientation="vertical"
      data-state="closed"
      hidden=""
      id="radix-:r1:"
      role="region"
      style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
    />
  </div>
</div>
`;

exports[`<ResultItem /> > should render result item correctly 1`] = `
<div
  class="space-y-2"
>
  <div
    class="rounded-lg border p-4"
  >
    <div
      class="flex items-center space-x-2"
    >
      <a
        href="https://github.com/farisphp/github-repo-explorer"
      >
        <h3
          class="w-full border-b text-base font-semibold hover:border-black"
        >
          github-repo-explorer
        </h3>
      </a>
      <div
        class="flex items-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
        <span
          class="ml-1"
        >
          0
        </span>
      </div>
      <span
        class="rounded-md border p-1 text-sm font-light"
      >
        TypeScript
      </span>
    </div>
    <span
      class="mt-1 block"
    >
      Last update 
      about 2 hours
    </span>
    <p
      class="mt-2"
    >
      Find a user and their public repositories by their username.
    </p>
  </div>
</div>
`;
